packages/python/plotly/plotly/express/_imshow.py

```diff
@@ -0,0 +1,122 @@
+import plotly.graph_objs as go
+import numpy as np  # is it fine to depend on np here?
+
+_float_types = []
+
+# Adapted from skimage.util.dtype
+_integer_types = (
+    np.byte,
+    np.ubyte,  # 8 bits
+    np.short,
+    np.ushort,  # 16 bits
+    np.intc,
+    np.uintc,  # 16 or 32 or 64 bits
+    np.int_,
+    np.uint,  # 32 or 64 bits
+    np.longlong,
+    np.ulonglong,
+)  # 64 bits
+_integer_ranges = {t: (np.iinfo(t).min, np.iinfo(t).max) for t in _integer_types}
+
+
+def _vectorize_zvalue(z):
+    if z is None:
+        return z
+    elif np.isscalar(z):
+        return [z] * 3 + [1]
+    elif len(z) == 1:
+        return list(z) * 3 + [1]
+    elif len(z) == 3:
+        return list(z) + [1]
+    elif len(z) == 4:
+        return z
+    else:
+        raise ValueError(
+            "zmax can be a scalar, or an iterable of length 1, 3 or 4. "
+            "A value of %s was passed for zmax." % str(z)
+        )
+
+
+def _infer_zmax_from_type(img):
+    dt = img.dtype.type
+    if dt in _integer_types:
+        return _integer_ranges[dt][1]
+    else:
+        return img[np.isfinite(img)].max()
```

Comment:
Per our conversation, I think this should return "the smallest power of 255 which is greater than the max".

My rationale here is that if you have a pipeline that's intended to produce output between 0-X, and you use this to display the output of multiple inputs, it would be nice for them to have the same bounds, instead of having the bounds vary. And it feels like the most likely values of X are 1 and 255, and then possibly thereafter some powers thereof, if the data is 16-bit or 32-bit or whatnot :)

Reply:
one thing which bothers me though is the following case: suppose an image is in the [0-1] range but some numerical computation (filter etc.) changes the max to 1 + some small value. Then the zmax will be 255 for a max of say 1.05, and it will look really bad.

Reply:
I can add a multiplicative tolerance factor like 10%

Reply:
I'd be OK with this

Reply:
how about having a keyword argument `colorrange` (or a better name consistent maybe with other parts of the library) which could be `data` (take min/max) or `dtype` (inference based on what we've discussed)? With this I think most people would be happy... Or does this complicate too much the API ?

Reply:
My feeling is that that complicates too much